include/lsst/afw/image/PhotoCalib.h

```diff
@@ -0,0 +1,426 @@
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+#ifndef LSST_AFW_IMAGE_PHOTOCALIB_H
+#define LSST_AFW_IMAGE_PHOTOCALIB_H
+
+/**
+ * The photometric calibration of an exposure.
+ *
+ * A PhotoCalib is a BoundedField (a function with a specified domain) that converts between calibrated
+ * counts-on-chip (ADU) to flux and magnitude. It is defined in terms of "maggies", which are a linear
+ * unit defined in SDSS: http://www.sdss.org/dr12/algorithms/magnitudes/#nmgy
+ *
+ * PhotoCalib is immutable.
+ *
+ * The spatially varying flux/magnitude zero point is defined such that,
+ * at a position (x,y) in the domain of the boundedField zeroPoint
+ * and for a given measured source counts:
+ * @f[
+ *     counts / zeroPoint(x,y) = flux (in maggies)
+ * @f]
+ * while the errors (constant on the domain) are defined as:
+ * @f[
+ *     sqrt((countsSigma/counts)^2 + (zeroPointSigma/zeroPoint)^2)*flux = fluxSigma (in maggies)
+ * @f]
+ * This implies that the conversions from counts and counts error to magnitude and magnitude error
+ * are as follows:
+ * @f[
+ *     -2.5 * log_10(counts / zeroPoint(x,y)) = magnitude
+ * @f]
+ * and
+ * @f[
+ *     2.5*log(10) * sqrt((countsSigma/counts)^2 + (zeroPointSigma/zeroPoint)^2) = magSigma
+ * @f]
+ */
+
+#include "lsst/afw/math/BoundedField.h"
+#include "lsst/afw/math/ChebyshevBoundedField.h"
+#include "lsst/afw/geom/Point.h"
+#include "lsst/afw/geom/Box.h"
+#include "lsst/afw/table/Source.h"
+#include "lsst/afw/table/io/Persistable.h"
+
+namespace lsst { namespace afw { namespace image {
+
+/// A value and its error.
+struct Quantity {
+    Quantity(double value, double sigma): value(value), sigma(sigma) {}
+    double const value;
+    double const sigma;
+};
+
+class PhotoCalib : public table::io::PersistableFacade<PhotoCalib>, public table::io::Persistable {
+public:
+    // no move or copy
+    PhotoCalib(PhotoCalib const &) = delete;
+    PhotoCalib(PhotoCalib &&) = delete;
+    PhotoCalib & operator=(PhotoCalib const &) = delete;
+    PhotoCalib & operator=(PhotoCalib &&) = delete;
+
+    /**
+     * Create a empty, zeroed calibration.
+     */
+    PhotoCalib(): PhotoCalib(0) {}
+
+    /**
+     * Create a non-spatially-varying calibration.
+     *
+     * @param[in]  countsMag0       The constant counts/magnitude zero point (counts at magnitude 0).
+     * @param[in]  countsMag0Sigma  The error on the zero point.
+     * @param[in]  bbox             The bounding box on which this PhotoCalib is valid.
+     */
+    explicit PhotoCalib(double countsMag0, double countsMag0Sigma=0,
+                        afw::geom::Box2I const & bbox=afw::geom::Box2I()):
+        _countsMag0(countsMag0), _countsMag0Sigma(countsMag0Sigma), _isConstant(true) {
+            ndarray::Array<double, 2, 2> coeffs = ndarray::allocate(ndarray::makeVector(1, 1));
+            coeffs[0][0] = countsMag0;
+            _zeroPoint = std::make_shared<afw::math::ChebyshevBoundedField>(
```

Comment:
It's about a factor of 2, in my testing. The time appears to be mostly independent of the complexity of the BoundedField, so it's likely due to the lookup of the centroids.

There's more to it than that, though: a constant PhotoCalib can be created without a bounding box, in which case an "empty" bbox gets created. Empty bboxes have some strange properties, which one would have to work around separately. By treating known-constant PhotoCalibs separately, we can avoid that issue, while keeping the empty bbox.